src/hdmf/common/resources.py

```diff
@@ -512,12 +551,58 @@ def add_ref(self, **kwargs):
                     msg = "Use Key Object when referencing an existing (container, relative_path, key)"
                     raise ValueError(msg)
 
-        if not isinstance(key, Key):
             key = self._add_key(key)
             self._add_object_key(object_field, key)
 
-        entity = self._add_entity(key, entity_id, entity_uri)
+        else:
+            # Check to see that the existing key is being used with the object.
+            # If true, do nothing. If false, create a new obj/key relationship
+            # in the ObjectKeyTable
+            key_idx = key.idx
+            object_key_row_idx = self.object_keys.which(keys_idx=key_idx)
+            if len(object_key_row_idx)!=0:
+                obj_key_check = False
+                for row_idx in object_key_row_idx:
+                    obj_idx = self.object_keys['objects_idx', row_idx]
+                    if obj_idx == object_field.idx:
+                        obj_key_check = True
+                if not obj_key_check:
+                    self._add_object_key(object_field, key)
+            else:
+                msg = "Cannot find key object. Create new Key with string."
+                raise ValueError(msg)
+            # check if the key and object have been related in the ObjectKeyTable
 
+        entity = self.get_entity(entity_id=entity_id)
+        if isinstance(entity, bool):
```

Comment:
```suggestion
        if entity is None:
```
Part of the suggested change above.